File: frontend/src/api/modelCatalog.ts
import { apiClient } from './client'

export interface ModelCatalogCategory {
  id: string
  label: string
  description?: string
  models: string[]
}

export interface ModelCatalogConfig {
  categories: ModelCatalogCategory[]
}

export const DEFAULT_MODEL_CATALOG: ModelCatalogConfig = {
  categories: [
    {
      id: 'openai',
      label: 'OpenAI',
      description: '可用模型列表。',
      models: [
        'gpt-5.2',
        'gpt-5.3-codex',
        'gpt-5.3-codex-spark',
        'gpt-5.4',
        'gpt-5.4-mini',
        'gpt-5.5',
        'gpt-image-2',
      ],
    },
  ],
}

function normalizeModelName(model: unknown): string | null {
  if (typeof model !== 'string') return null
  const trimmed = model.trim()
  return trimmed || null
}

export function normalizeModelCatalogConfig(input: unknown): ModelCatalogConfig {
  if (!input || typeof input !== 'object') return DEFAULT_MODEL_CATALOG
  const rawCategories = (input as { categories?: unknown }).categories
  if (!Array.isArray(rawCategories)) return DEFAULT_MODEL_CATALOG

  const categories = rawCategories
    .map((category, index): ModelCatalogCategory | null => {
      if (!category || typeof category !== 'object') return null
      const raw = category as Record<string, unknown>
      const models = Array.isArray(raw.models)
        ? Array.from(new Set(raw.models.map(normalizeModelName).filter(Boolean) as string[]))
        : []
      if (models.length === 0) return null
      const id = typeof raw.id === 'string' && raw.id.trim() ? raw.id.trim() : `category-${index + 1}`
      const label = typeof raw.label === 'string' && raw.label.trim() ? raw.label.trim() : id
      const description = typeof raw.description === 'string' ? raw.description.trim() : ''
      return { id, label, description, models }
    })
    .filter(Boolean) as ModelCatalogCategory[]

  return categories.length > 0 ? { categories } : DEFAULT_MODEL_CATALOG
}

export async function getPublicModelCatalog(): Promise<ModelCatalogConfig> {
  const { data } = await apiClient.get<ModelCatalogConfig>('/model-catalog')
  return normalizeModelCatalogConfig(data)
}

export async function getAdminModelCatalog(): Promise<ModelCatalogConfig> {
  const { data } = await apiClient.get<ModelCatalogConfig>('/admin/model-catalog')
  return normalizeModelCatalogConfig(data)
}

export async function updateAdminModelCatalog(config: ModelCatalogConfig): Promise<ModelCatalogConfig> {
  const { data } = await apiClient.put<ModelCatalogConfig>('/admin/model-catalog', config)
  return data
}

File: backend/internal/service/model_catalog.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
)

type ModelCatalogConfig struct {
	Categories []ModelCatalogCategory `json:"categories"`
}

type ModelCatalogCategory struct {
	ID          string   `json:"id"`
	Label       string   `json:"label"`
	Description string   `json:"description,omitempty"`
	Models      []string `json:"models"`
}

var defaultModelCatalogConfig = ModelCatalogConfig{Categories: []ModelCatalogCategory{
	{
		ID:          "openai",
		Label:       "OpenAI 兼容 / 奥特曼加",
		Description: "本站当前主推的奥特曼加模型，适配 OpenAI 兼容接口。",
		Models: []string{
			"gpt-5.2",
			"gpt-5.3-codex",
			"gpt-5.3-codex-spark",
			"gpt-5.4",
			"gpt-5.4-mini",
			"gpt-5.5",
			"gpt-image-2",
		},
	},
}}

func DefaultModelCatalogConfig() ModelCatalogConfig {
	return defaultModelCatalogConfig
}

func NormalizeModelCatalogConfig(config ModelCatalogConfig) ModelCatalogConfig {
	categories := make([]ModelCatalogCategory, 0, len(config.Categories))
	for i, category := range config.Categories {
		models := make([]string, 0, len(category.Models))
		seen := make(map[string]struct{}, len(category.Models))
		for _, model := range category.Models {
			model = strings.TrimSpace(model)
			if model == "" {
				continue
			}
			if _, ok := seen[model]; ok {
				continue
			}
			seen[model] = struct{}{}
			models = append(models, model)
		}
		if len(models) == 0 {
			continue
		}
		id := strings.TrimSpace(category.ID)
		if id == "" {
			id = fmt.Sprintf("category-%d", i+1)
		}
		label := strings.TrimSpace(category.Label)
		if label == "" {
			label = id
		}
		categories = append(categories, ModelCatalogCategory{
			ID:          id,
			Label:       label,
			Description: strings.TrimSpace(category.Description),
			Models:      models,
		})
	}
	if len(categories) == 0 {
		return DefaultModelCatalogConfig()
	}
	return ModelCatalogConfig{Categories: categories}
}

func (s *SettingService) GetModelCatalog(ctx context.Context) (ModelCatalogConfig, error) {
	raw, err := s.settingRepo.GetValue(ctx, SettingKeyModelCatalog)
	if err != nil || strings.TrimSpace(raw) == "" {
		return DefaultModelCatalogConfig(), nil
	}
	var config ModelCatalogConfig
	if err := json.Unmarshal([]byte(raw), &config); err != nil {
		return DefaultModelCatalogConfig(), nil
	}
	return NormalizeModelCatalogConfig(config), nil
}

func (s *SettingService) UpdateModelCatalog(ctx context.Context, config ModelCatalogConfig) (ModelCatalogConfig, error) {
	normalized := NormalizeModelCatalogConfig(config)
	payload, err := json.Marshal(normalized)
	if err != nil {
		return ModelCatalogConfig{}, err
	}
	if err := s.settingRepo.Set(ctx, SettingKeyModelCatalog, string(payload)); err != nil {
		return ModelCatalogConfig{}, err
	}
	return normalized, nil
}
